scripts/who-uploads.1
.TH WHO-UPLOADS 1 "Debian Utilities" "DEBIAN" \" -*- nroff -*-
.SH NAME
who-uploads \- identify the uploaders of Debian source packages
.SH SYNOPSIS
\fBwho\-uploads\fR [\fIoptions\fR] \fIsource_package\fR ...
.SH DESCRIPTION
\fBwho\-uploads\fR uses the Package Tracking System (PTS) to identify
the uploaders of the three most recent versions of the given source
packages.  Note that the uploaders are identified using their
\fBgpg\fR(1) or \fBgpg2\fR(1) keys; installing a recent version of the
\fIdebian-keyring\fR package should provide most of the required keys.
.PP
Note that because the PTS uses source packages, you must give the
source package names, not the binary package names.
.SH OPTIONS
.TP
\fB\-M\fR, \fB\-\-max\-uploads=\fIN\fR
Specify the maximum number of uploads to display for each package; the
default is 3.  Note that you may not see this many uploads if there
are not this many recorded in the PTS.
.TP
\fB\-\-keyring \fIkeyring\fR
Add \fIkeyring\fR to the list of keyrings to be searched for the
uploader's GPG key.
.TP
\fB\-\-no\-default\-keyrings\fR
By default, \fBwho\-uploads\fR uses the three Debian keyrings
\fI/usr/share/keyrings/debian-keyring.gpg\fR,
\fI/usr/share/keyrings/debian-nonupload.gpg\fR, and
\fI/usr/share/keyrings/debian-maintainers.gpg\fR (although this
default can be changed in the configuration file, see below).
Specifying this option means that the default keyrings will not be
examined.  The \fB\-\-keyring\fR option overrides this one.
.TP
\fB\-\-date\fR
Show the date of the upload alongside the uploader's details
.TP
.BR \-\-nodate ", " \-\-no\-date
Do not show the date of the upload alongside the uploader's details.
This is the default behaviour.
.TP
\fB\-\-no-conf\fR, \fB\-\-noconf\fR
Do not read any configuration files.  This can only be used as the
first option given on the command-line.
.TP
.BR \-\-help ", " \-h
Display a help message and exit successfully.
.TP
.B \-\-version
Display version and copyright information and exit successfully.
.SH "CONFIGURATION VARIABLES"
The two configuration files \fI/etc/devscripts.conf\fR and
\fI~/.devscripts\fR are sourced in that order to set configuration
variables.  Command line options can be used to override configuration
file settings.  Environment variable settings are ignored for this
purpose.  The currently recognised variables are:
.TP
.B WHOUPLOADS_DATE
Show the date of the upload alongside the uploader's details.  By
default, this is "no".
.TP
.B WHOUPLOADS_MAXUPLOADS
The maximum number of uploads to display for each package.  By
default, this is 3.
.TP
.B WHOUPLOADS_KEYRINGS
This is a colon-separated list of the default keyrings to be used.  By
default, it is the three Debian keyrings
\fI/usr/share/keyrings/debian-keyring.gpg\fR,
\fI/usr/share/keyrings/debian-nonupload.gpg\fR,
and
\fI/usr/share/keyrings/debian-maintainers.gpg\fR.
.SH AUTHOR
The original version of \fBwho-uploads\fR was written by Adeodato Sim\['o]
<dato@net.com.org.es>.  The current version is by Julian Gilbey
<jdg@debian.org>.


debian/NEWS
devscripts (2.16.10) unstable; urgency=medium

  debuild's target mode (i.e. where it runs the specified debian/rules
  targets) now requires a '--' between the debuild options and the targets
  that need to be run.

  For example, "debuild -d -- build-arch binary-arch".

 -- James McCoy <jamessan@debian.org>  Thu, 24 Nov 2016 14:25:12 -0500

devscripts (2.14.0) unstable; urgency=medium

  uscan now looks for upstream's keyring under debian/upstream/ as either
  signing-key.pgp or, for an armored keyring, signing-key.asc.  The previously
  used location, debian/upstream-signing-key.pgp, will be checked as a last
  resort for a transition period.

 -- James McCoy <jamessan@debian.org>  Sat, 25 Jan 2014 10:19:02 -0500

devscripts (2.11.9) unstable; urgency=low

  The default for mk-build-deps --tool option/MKBUILDDEPS_TOOL configuration
  changed from "apt-get" to "apt-get --no-install-recommends".

 -- James McCoy <jamessan@debian.org>  Sun, 17 Jun 2012 22:56:26 +0200

devscripts (2.11.7) unstable; urgency=low

  The heuristic used to determine what action to perform when no options are
  given to debchange now defaults to 'changelog'.  To stay with the previous
  behavior, DEBCHANGE_RELEASE_HEURISTIC=log can be added to the configuration
  file.

  Enabling the changelog heuristic also enables DEBCHANGE_MAINTTRAILER.  This
  can be disabled by adding DEBCHANGE_MAINTTRAILER=no to the configuration
  file.

 -- James McCoy <jamessan@debian.org>  Sun, 29 Apr 2012 18:38:46 -0400

devscripts (2.11.2) unstable; urgency=low

  Support for pgp-style commands has been removed from debsign/debrsign.  As
  such, the DEBSIGN_SIGNLIKE configuration variable has also been removed.

 -- James McCoy <jamessan@debian.org>  Tue, 15 Nov 2011 21:36:40 -0500

devscripts (2.10.54) unstable; urgency=low

    manpage-alert now, as originally intended, prints out the full path to the
    binary which is lacking a manpage instead of simply the binary name.

    Before:
      No manual entry for somebinary

    After:
      No manual entry for /usr/bin/somebinary

 -- James Vega <jamessan@debian.org>  Thu, 06 Aug 2009 18:03:41 -0400

devscripts (2.10.24) unstable; urgency=low

    Starting with devscripts 2.10.21, debchange attempts to
    automatically determine whether an NMU changelog should be
    started, by comparing DEBEMAIL and DEBFULLNAME to the
    package's Maintainer and Uploaders fields.

    This behaviour may now be disabled using the DEBCHANGE_AUTO_NMU
    configuration file variable.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Mon, 07 Apr 2008 14:38:08 +0100

devscripts (2.10.20) unstable; urgency=low

    devscripts 2.10.19 modified debcommit's --release option to use the
    uncommitted changes (if any) in the changelog to derive the commit
    message.

    This behaviour is now optional and can be enabled using the -R /
    --release-use-changelog command line option or by setting the
    DEBCOMMIT_RELEASE_USE_CHANGELOG configuration option to "yes".

    The default is to use the previous (i.e. pre-2.10.19) behaviour.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Wed, 19 Mar 2008 20:31:20 +0000

devscripts (2.10.19) unstable; urgency=low

    In devscripts 2.10.17, dget was modified to unpack source packages by
    default.  The new -d switch or the DGET_UNPACK configuration file option
    may be used to indicate that the unpacking should not occur.

    As linda has been removed from both unstable and testing, support for
    it has been removed from debuild and cvs-debuild.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Thu, 13 Mar 2008 19:33:38 +0000


scripts/who-uploads.sh
#!/bin/bash

# who-uploads sourcepkg [ sourcepkg ... ]
# Tells you who made the latest uploads of a source package.
# NB: I'm encoded in UTF-8!!

# Written and copyright 2006 by Julian Gilbey <jdg@debian.org>
# Based on an original script
# copyright 2006 Adeodato Simó <dato@net.com.org.es>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

set -e

PROGNAME=`basename $0`
MODIFIED_CONF_MSG='Default settings modified by devscripts configuration files:'

usage () {
    echo \
"Usage: $PROGNAME [options] package ...
  Display the most recent three uploaders of each package.
  Packages should be source packages, not binary packages.

  Options:
    -M, --max-uploads=N
                      Display at most the N most recent uploads (default: 3)
    --keyring KEYRING Add KEYRING as a GPG keyring for Debian Developers'
                      keys in addition to /usr/share/keyrings/debian-keyring.*
                      and /usr/share/keyrings/debian-maintainers.gpg;
                      this option may be given multiple times
    --no-default-keyrings
                      Do not use the default keyrings
    --no-conf, --noconf
                      Don't read devscripts config files;
                      must be the first option given
    --date            Display the date of the upload
    --no-date, --nodate
                      Don't display the date of the upload (default)
    --help            Show this message
    --version         Show version and copyright information

$MODIFIED_CONF_MSG"
}

version () {
    echo \
"This is $PROGNAME, from the Debian devscripts package, version ###VERSION###
This code is copyright 2006 by Julian Gilbey <jdg@debian.org>,
all rights reserved.
Based on original code copyright 2006 Adeodato Simó <dato@net.com.org.es>
This program comes with ABSOLUTELY NO WARRANTY.
You are free to redistribute this code under the terms of the
GNU General Public License, version 2 or later."
}


# Boilerplate: set config variables
DEFAULT_WHOUPLOADS_KEYRINGS=/usr/share/keyrings/debian-keyring.gpg:/usr/share/keyrings/debian-maintainers.gpg
DEFAULT_WHOUPLOADS_MAXUPLOADS=3
DEFAULT_WHOUPLOADS_DATE=no
VARS="WHOUPLOADS_KEYRINGS WHOUPLOADS_MAXUPLOADS WHOUPLOADS_DATE"

GPG=gpg
if ! command -v $GPG >/dev/null 2>&1; then
    echo "$GPG missing"
    GPG=gpg2
    if  ! command -v $GPG >/dev/null 2>&1; then
	echo "$GPG missing"
	exit 1
    fi
fi

if [ "$1" = "--no-conf" -o "$1" = "--noconf" ]; then
    shift
    MODIFIED_CONF_MSG="$MODIFIED_CONF_MSG
  (no configuration files read)"

    # set defaults
    for var in $VARS; do
	eval "$var=\$DEFAULT_$var"
    done
else
    # Run in a subshell for protection against accidental errors
    # in the config files
    eval $(
	set +e
	for var in $VARS; do
	    eval "$var=\$DEFAULT_$var"
	done

	for file in /etc/devscripts.conf ~/.devscripts
	  do
	  [ -r $file ] && . $file
	done

	set | grep "^WHOUPLOADS_")

    # check sanity
    if [ "$WHOUPLOADS_MAXUPLOADS" != \
	    "$(echo \"$WHOUPLOADS_MAXUPLOADS\" | tr -cd 0-9)" ]; then
	WHOUPLOADS_MAXUPLOADS=3
    fi

    WHOUPLOADS_DATE="$(echo "$WHOUPLOADS_DATE" | tr A-Z a-z)"
    if [ "$WHOUPLOADS_DATE" != "yes" ] && [ "$WHOUPLOADS_DATE" != "no" ]; then
	WHOUPLOADS_DATE=no
    fi

    # don't check WHOUPLOADS_KEYRINGS here

    # set config message
    MODIFIED_CONF=''
    for var in $VARS; do
	eval "if [ \"\$$var\" != \"\$DEFAULT_$var\" ]; then
	    MODIFIED_CONF_MSG=\"\$MODIFIED_CONF_MSG
  $var=\$$var\";
	MODIFIED_CONF=yes;
	fi"
    done

    if [ -z "$MODIFIED_CONF" ]; then
	MODIFIED_CONF_MSG="$MODIFIED_CONF_MSG
  (none)"
    fi
fi

MAXUPLOADS=$WHOUPLOADS_MAXUPLOADS
WANT_DATE=$WHOUPLOADS_DATE

declare -a GPG_DEFAULT_KEYRINGS
declare -a GPG_KEYRINGS

# Command-line options
TEMP=$(getopt -s bash -o 'h' \
	--long max-uploads:,keyring:,no-default-keyrings \
	--long no-conf,noconf \
	--long date,nodate,no-date \
	--long help,version \
	--options M: \
	-n "$PROGNAME" -- "$@") || (usage >&2; exit 1)

eval set -- $TEMP

# Process Parameters
while [ "$1" ]; do
    case $1 in
    --max-uploads|-M)
	shift
	if [ "$1" = "$(echo \"$1\" | tr -cd 0-9)" ]; then
	    MAXUPLOADS=$1
	fi
	;;
    --keyring)
	shift
	if [ -f "$1" ]; then
	    GPG_KEYRINGS=("${GPG_KEYRINGS[@]}" "--keyring" "$1")
	else
	    echo "Could not find keyring $1, skipping" >&2
	fi
	;;
    --no-default-keyrings)
	WHOUPLOADS_KEYRINGS=
	;;
    --no-conf|--noconf)
	echo "$PROGNAME: $1 is only acceptable as the first command-line option!" >&2
	exit 1 ;;
    --date) WANT_DATE=yes ;;
    --no-date|--nodate) WANT_DATE=no ;;
    --help|-h) usage; exit 0 ;;
    --version) version; exit 0 ;;
    --)	shift; break ;;
    *) echo "$PROGNAME: bug in option parser, sorry!" >&2 ; exit 1 ;;
    esac
    shift
done

OIFS="$IFS"
IFS=:

for keyring in $WHOUPLOADS_KEYRINGS; do
    if [ -f "$keyring" ]; then
	GPG_DEFAULT_KEYRINGS=("${GPG_DEFAULT_KEYRINGS[@]}" "--keyring" "$keyring")
    elif [ -n "$keyring" ]; then
	echo "Could not find keyring $keyring, skipping it" >&2
    fi
done

IFS="${OIFS:- 	}"

GNUPGHOME=$(mktemp -d)
exit_with_error() {
    [ ! -d "$GNUPGHOME" ] || rm -r "$GNUPGHOME"
    exit 1
}
trap exit_with_error HUP INT QUIT PIPE ALRM TERM
export GNUPGHOME

# Some useful abbreviations for gpg options
GPG_OPTIONS="--no-options --no-auto-check-trustdb --no-default-keyring"
GPG_NO_KEYRING="$GPG_OPTIONS --keyring /dev/null"

if [ $# -eq 0 ]; then
    usage;
    exit 1
fi

# Now actually get the reports :)

for package; do
    echo "Uploads for $package:"

    prefix=$(echo $package | sed -re 's/^((lib)?.).*$/\1/')
    pkgurl="https://packages.qa.debian.org/${prefix}/${package}.html"
    baseurl="https://packages.qa.debian.org/${prefix}/"

    # only grab the actual "Accepted" news announcements; hopefully this
    # won't pick up many false positives
    WGETOPTS="-q -O - --timeout=30 "
    count=0
    for news in $(wget $WGETOPTS $pkgurl |
                  sed -ne 's%^.*<a href="\('$package'/news/[0-9A-Z]*\.html\)">Accepted .*%\1%p'); do
	HTML_TEXT=$(wget $WGETOPTS "$baseurl$news")
	GPG_TEXT=$(echo "$HTML_TEXT" |
	           sed -ne 's/^<pre>//; /-----BEGIN PGP SIGNED MESSAGE-----/,/-----END PGP SIGNATURE-----/p')

	test -n "$GPG_TEXT" || continue

	VERSION=$(echo "$GPG_TEXT" | awk '/^Version/ { print $2; exit }')
	DISTRO=$(echo "$GPG_TEXT" | awk '/^Distribution/ { print $2; exit }')
	if [ "$WANT_DATE" = "yes" ]; then
	    DATE=$(echo "$HTML_TEXT" |  sed -ne 's%<li><em>Date</em>: \(.*\)</li>%\1%p')
	fi

	GPG_ID=$(echo "$GPG_TEXT" | LC_ALL=C $GPG $GPG_NO_KEYRING --keyid-format long --verify 2>&1 |
	         sed -rne 's/.*using [^ ]* key ([0-9A-Z]+).*/\1/p')

	UPLOADER=$($GPG $GPG_OPTIONS \
	           "${GPG_DEFAULT_KEYRINGS[@]}" "${GPG_KEYRINGS[@]}" \
	           --list-key --with-colons --fixed-list-mode $GPG_ID 2>/dev/null |
	           awk  -F: '/@debian\.org>/ { a = $10; exit} /^uid/ { a = $10; exit} END { print a }' )
	if [ -z "$UPLOADER" ]; then UPLOADER="<unrecognised public key ($GPG_ID)>"; fi

	output="$VERSION to $DISTRO: $UPLOADER"
	[ "$WANT_DATE" = "yes" ] && output="$output on $DATE"
	echo $output | iconv -c -f UTF-8

	count=$(($count + 1))
	[ $count -eq $MAXUPLOADS ] && break
    done
    test $# -eq 1 || echo
done

[ ! -d "$GNUPGHOME" ] || rm -r "$GNUPGHOME"
exit 0
